zcash_client_sqlite/src/wallet.rs

```diff
@@ -461,30 +515,78 @@ pub fn rewind_to_height<P: consensus::Parameters>(
                     .or(Ok(sapling_activation_height - 1))
             })?;
 
-    // nothing to do if we're deleting back down to the max height
-    if block_height >= last_scanned_height {
-        Ok(())
+    let rewind_height = if block_height >= (last_scanned_height - PRUNING_HEIGHT) {
+        Some(block_height)
     } else {
-        // Decrement witnesses.
-        wdb.conn.execute(
-            "DELETE FROM sapling_witnesses WHERE block > ?",
-            &[u32::from(block_height)],
-        )?;
-
-        // Un-mine transactions.
-        wdb.conn.execute(
-            "UPDATE transactions SET block = NULL, tx_index = NULL WHERE block > ?",
-            &[u32::from(block_height)],
-        )?;
+        match get_rewind_height(&wdb)? {
+            Some(h) => {
+                if block_height > h {
+                    return Err(SqliteClientError::RequestedRewindInvalid(h));
+                } else {
+                    Some(block_height)
+                }
+            }
+            None => Some(block_height),
+        }
+    };
 
-        // Now that they aren't depended on, delete scanned blocks.
-        wdb.conn.execute(
-            "DELETE FROM blocks WHERE height > ?",
-            &[u32::from(block_height)],
-        )?;
+    // nothing to do if we're deleting back down to the max height
 
-        Ok(())
+    if let Some(block_height) = rewind_height {
```

Comment:
This is unnecessary; `block_height` is always Some. (There's a return for the `RequestedRewindInvalid` case, but the block above never evaluates to None.) Simplify by removing this `if let`.